File: macos/Sources/ElevateCore/Coordination/ProfilePlanner.swift
import Foundation

/// One line of a profile run: what will happen to an entry and with which duration.
public struct ProfilePlanItem: Hashable, Sendable, Identifiable {
    public enum Disposition: Hashable, Sendable { case activate, alreadyActive, pending, notEligible }
    public let roleKey: RoleKey
    public let role: EligibleRole?
    public var duration: Duration
    public let disposition: Disposition
    public var id: RoleKey { roleKey }
    public init(roleKey: RoleKey, role: EligibleRole?, duration: Duration, disposition: Disposition) {
        self.roleKey = roleKey; self.role = role; self.duration = duration; self.disposition = disposition
    }
}

public enum ProfilePlanner {
    /// Duration: the entry's last run, else the role's remembered duration, else the policy default; never above the maximum.
    public static func plan(_ profile: ActivationProfile, roles: [RoleKey: EligibleRole],
                            active: [RoleKey: ActiveAssignment], memory: [RoleKey: RoleMemory]) -> [ProfilePlanItem] {
        profile.entries.map { entry in
            let role = roles[entry.roleKey]
            let policy = role?.policy ?? .manualDefault
            let wanted = entry.lastDuration ?? memory[entry.roleKey]?.lastDuration ?? policy.defaultDuration
            let duration = min(wanted, policy.maximumDuration)
            let disposition: ProfilePlanItem.Disposition
            if role == nil { disposition = .notEligible }
            else if let a = active[entry.roleKey] {
                switch a.status {
                case .active: disposition = .alreadyActive
                case .pendingApproval, .pendingProvisioning: disposition = .pending
                case .failed: disposition = .activate
                }
            } else { disposition = .activate }
            return ProfilePlanItem(roleKey: entry.roleKey, role: role, duration: duration, disposition: disposition)
        }
    }
}

File: windows/src/Elevate.Core/Coordination/ProfilePlanner.cs
using Elevate.Core.Models;
using Elevate.Core.Storage;

namespace Elevate.Core.Coordination;

/// <summary>What a profile run will do with one entry. Port of Swift's <c>ProfilePlanItem.Disposition</c>.</summary>
public enum ProfilePlanDisposition { Activate, AlreadyActive, Pending, NotEligible, NotLoaded }

/// <summary>One line of a profile run: what will happen to an entry and with which duration.</summary>
public sealed record ProfilePlanItem(RoleKey RoleKey, EligibleRole? Role, TimeSpan Duration, ProfilePlanDisposition Disposition);

/// <summary>Port of the Swift <c>ProfilePlanner</c>.</summary>
public static class ProfilePlanner
{
    /// <summary>
    /// Duration: the entry's last run, else the role's remembered duration, else the policy default; never above the maximum.
    /// <paramref name="loadedTenants"/> are the tenants whose roles are known; a missing role in any other tenant
    /// is "not loaded yet", not "not eligible".
    /// </summary>
    public static IReadOnlyList<ProfilePlanItem> Plan(
        ActivationProfile profile,
        IReadOnlyDictionary<RoleKey, EligibleRole> roles,
        IReadOnlyDictionary<RoleKey, ActiveAssignment> active,
        IReadOnlyDictionary<RoleKey, RoleMemory> memory,
        IReadOnlySet<TenantKey> loadedTenants)
    {
        ArgumentNullException.ThrowIfNull(profile);
        ArgumentNullException.ThrowIfNull(roles);
        ArgumentNullException.ThrowIfNull(active);
        ArgumentNullException.ThrowIfNull(memory);
        ArgumentNullException.ThrowIfNull(loadedTenants);

        return [.. profile.Entries.Select(entry =>
        {
            var role = roles.GetValueOrDefault(entry.RoleKey);
            var policy = role?.Policy ?? RolePolicy.ManualDefault;
            var wanted = entry.LastDuration ?? memory.GetValueOrDefault(entry.RoleKey)?.LastDuration ?? policy.DefaultDuration;
            var duration = wanted < policy.MaximumDuration ? wanted : policy.MaximumDuration;
            ProfilePlanDisposition disposition;
            if (role is null)
            {
                disposition = loadedTenants.Contains(entry.RoleKey.TenantKey)
                    ? ProfilePlanDisposition.NotEligible
                    : ProfilePlanDisposition.NotLoaded;
            }
            else if (active.TryGetValue(entry.RoleKey, out var assignment))
            {
                disposition = assignment.Status.Kind switch
                {
                    AssignmentStatusKind.Active => ProfilePlanDisposition.AlreadyActive,
                    AssignmentStatusKind.PendingApproval or AssignmentStatusKind.PendingProvisioning or AssignmentStatusKind.Scheduled
                        => ProfilePlanDisposition.Pending,
                    _ => ProfilePlanDisposition.Activate,
                };
            }
            else
            {
                disposition = ProfilePlanDisposition.Activate;
            }

            return new ProfilePlanItem(entry.RoleKey, role, duration, disposition);
        })];
    }
}
